Add flattener post-processor for NewtonianEuler limiting

fmahebert opened the pull request:
## Proposed changes

Limiters try to fix oscillations that produce unphysical solution values like negative densities or pressures. But sometimes the algorithms fail to do so, and a further correction step is needed.

This PR adds a "flattener" or "bound-preserving filter" that scales the solution about its mean (u -> ubar + alpha * (u - ubar) for some alpha < 1). This brings every value closer to the mean, and can be used to remove points with negative densities and pressures.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

fmahebert committed:
Add flattener post-processor for NewtonianEuler limiters

nilsdeppe commented on src/Evolution/Systems/NewtonianEuler/Limiters/Flattener.hpp:
```diff
@@ -0,0 +1,58 @@
+/// After checking for (and correcting) negative densities, if the equation of
+/// state is two-dimensional, then the pressure is also checked for positivity.
+/// If negative pressures are found, each solution component is set to its mean
+/// (this is equivalent to \f$theta = 0\f$ in the scaling form above).
```
missing `\` in front of `theta`

nilsdeppe commented on src/Evolution/Systems/NewtonianEuler/Limiters/Flattener.cpp:
```diff
@@ -0,0 +1,152 @@
+    const Scalar<DataVector>& det_logical_to_inertial_jacobian,
+    const EquationsOfState::EquationOfState<false, ThermodynamicDim>&
+        equation_of_state) noexcept {
+  // Note on the design behind the handling of the cell-averaged fields:
```
I would find this easier to read as `A note on...` It took me a couple takes

nilsdeppe commented on src/Evolution/Systems/NewtonianEuler/Limiters/Flattener.cpp:
```diff
@@ -0,0 +1,152 @@
+    // compute the means
+    const double volume_of_cell =
+        definite_integral(get(det_logical_to_inertial_jacobian), mesh);
+    const auto jacobian_weighted_mean =
```
Isn't this just the mean? As in, you are integrating

```
int d^3\xi J u = int d^3x u
```
and then dividing by the "physical" volume.

nilsdeppe commented on src/Evolution/Systems/NewtonianEuler/Limiters/Flattener.cpp:
```diff
@@ -0,0 +1,152 @@
+    const auto jacobian_weighted_mean =
+        [&mesh, &det_logical_to_inertial_jacobian,
+         &volume_of_cell](const DataVector& u) noexcept {
+          return definite_integral(get(det_logical_to_inertial_jacobian) * u,
```
Just a heads up, I think you've got an allocation here because `definite_integral` is taking a `DataVector`.

fmahebert replied:
Sorry, I don't quite understand, can you clarify?

nilsdeppe replied:
The signature of definite integral is:
```
template <size_t Dim>
double definite_integral(const DataVector& integrand,
                         const Mesh<Dim>& mesh) noexcept;
```
so the expression `det_jac * u` does an allocation. Does that make sense?

fmahebert replied:
Understood, but I'm not sure what the alternative could be, because don't I need an allocation to store the product anyway? The only improvement I can think of would be to allocate a buffer outside the lambda and store the `det_jac * u` in there... it'd save me 4 (in 3d) allocations per call. Do you think I should do this?

nilsdeppe replied:
It likely won't make a difference, definitely not a big one. Maybe just add a comment to make folks aware? Expression templates make life confusing because of these subtle side effects :)

nilsdeppe commented on src/Evolution/Systems/NewtonianEuler/Limiters/Flattener.hpp:
```diff
@@ -0,0 +1,58 @@
+
+#include <cstddef>
+
+#include "DataStructures/Tensor/Tensor.hpp"
```
`TypeAliases.hpp` instead of `Tensor.hpp`

nilsdeppe commented on src/Evolution/Systems/NewtonianEuler/Limiters/Flattener.cpp:
```diff
@@ -0,0 +1,152 @@
+    }
+  };
+
+  size_t flattener_status = 0;
```
It seems this could be easier to read if it were an enum. It's not at all clear what the number returned means. I'd be happier with either a bool "yes I limited" or an enum "I limited this". If you need the return type to be convertible to a bool, I'd suggest an enum with explicitly specified values

nilsdeppe commented on tests/Unit/Evolution/Systems/NewtonianEuler/Limiters/Test_Flattener.cpp:
```diff
@@ -0,0 +1,378 @@
+namespace {
+
+void test_flattener_1d() noexcept {
+  const Mesh<1> mesh(4, Spectral::Basis::Legendre,
```
Might be good to add `INFO("Checking 1d");` or something along those lines

nilsdeppe commented on tests/Unit/Evolution/Systems/NewtonianEuler/Limiters/Test_Flattener.cpp:
```diff
@@ -0,0 +1,378 @@
+
+  // case with no flattening needed
+  {
+    Scalar<DataVector> mass_density_cons(DataVector{{0.8, 1.6, 2.2, 1.2}});
```
Might be good to add `INFO("Check no flattening needed");` instead of the c++ comment, similar below.

These changes are optional, just something to consider :)

nilsdeppe reviewed (changes requested):
I'm fine with you squashing any changes immediately

fmahebert commented:
Thanks for the review, @nilsdeppe! I've added a fixup commit because there were enough changes I thought it might make it easier for you to look things over. I've also rebased. Let me know if I've missed anything or if any of the comment changes can be further clarified.

nilsdeppe reviewed (commented):
LGTM! Please go ahead and squash :)

fmahebert commented:
Squashed. Thanks @nilsdeppe !

fmahebert commented:
@sxs-collaboration/spectre-core-devs can we move forward with this PR?

kidder commented on src/Evolution/Systems/NewtonianEuler/Limiters/Flattener.hpp:
```diff
@@ -0,0 +1,66 @@
+/// coordinates, so the flattener is conservative even on deformed grids.
+///
+/// A scaling of this form used to restore positivity is usually called a
+/// flattener (we use this name) or a bounds-preserving filters. Note that the
```
filters -> filter

kidder reviewed (commented):
you can squash immediately

fmahebert commented:
Rebased on recent develop, and squashed in the typo fix
